Add UPDATE, DELETE support, fix escaped quotes, detect invalid SQL

nerdrew opened the pull request:
cc @ggilder @maniksurtani @zanker

nerdrew commented:
Hmmm. I'm going to add some more commits to this. I'll break this into a bunch of smaller PRs.

Objections to a whitespace cleanup PR?

nerdrew committed:
large refactor

- add UPDATE
- add DELETE FROM
- add LIMIT
- add FOR UPDATE
- add DISTINCT
- add COALESCE function
- fix selecting literals, e.g. (1 + 1)
- refactor racc
- fix escaped quotes (both \' and '')
- better detection of invalid sql (e.g. SELECT1)

nerdrew commented:
Ok, punted on splitting into smaller PRs. They conflict too much. Is this change too large to review?

zanker commented:
I can review it, but I couldn't merge it so not sure if the repo owner might find it too big?

nerdrew commented:
@cryodex Do you have an opinion?

nerdrew commented:
@louismullie Thoughts on this PR?

nerdrew committed:
COALESCE, sql functions, 1E1

- fix COALESCE with multiple arguments
- add support for sql functions
- fix 1e1, 1E1 literals: don't match identifiers that start with "e"
